lib/application_insights/channel/contracts/data.rb

```diff
@@ -1,48 +1,9 @@
 require_relative 'json_serializable'
 
-module ApplicationInsights
-  module Channel
-    module Contracts
-      # Data contract class for type Data.
-      class Data < JsonSerializable
-        # Initializes a new instance of the Data class.
-        def initialize(options={})
-          defaults = {
-            'baseType' => nil,
-            'baseData' => nil
-          }
-          values = {
-            'baseData' => nil
-          }
-          super defaults, values, options
-        end
-        
-        # Gets the base_type property.
-        def base_type
-          @values.fetch('baseType') { 
-            @values['baseType'] = nil
-          }
-        end
-        
-        # Sets the base_type property.
-        def base_type=(value)
-          if value == @defaults['baseType']
-            @values.delete 'baseType' if @values.key? 'baseType'
-          else
-            @values['baseType'] = value
-          end
-        end
-        
-        # Gets the base_data property.
-        def base_data
-          @values['baseData']
-        end
-        
-        # Sets the base_data property.
-        def base_data=(value)
-          @values['baseData'] = value
-        end
-      end
-    end
+module ApplicationInsights::Channel::Contracts
+  class Data
+    include JsonSerializable
+
+    attributes :baseType, :baseData
```

Comment:
As mentioned before, the contract classes are generated so that it's easy to maintain with schema changes. And the generator not only used for ruby SDK but other AI SDKs as well.  For this reason, I don't quite consider the option of Example 3 which includes the explicit name transformation logic which should be better inside the generator.

•you must send a mapping hash in the  initialize  method
  [Jing] The mapping is implementation detail.   I'd like to avoid it to be part of interface if it's not necessary.
•you have to define the accessor only for those who has no default value
 [Jing] We could define accessor for all fields and override reads for those with default values.  However, I don't see a good reason to do method redefining as it's all controlled in one class. Also, this is easy to control in the generator when to use attr_accessor or attr_writer.
•you have to define the setters for the attributes with default values which are defined later.
[Jing] Yes, redefine the getter to get the default value when it's not set. Sounds a good idea to me the way you are doing in the current pull request.

The 'JsonSerializable' is more like a characteristic of telemetry data, in that case, I agree using mixins is better than inheritance.